Implement divide

SeanLF opened the pull request.

SeanLF committed:
Implement divide

SeanLF committed:
Wip Implement group

Not guaranteed to finish, but usually finds the black hole

SeanLF committed:
Small improvement & refactor

SeanLF committed:
As good as it gets

Can find black hole up to 3rd quarter with perfect accuracy

jonathanGB commented on bhs/agent.go:
```diff
@@ -206,6 +206,20 @@ func (agent *Agent) LeaveUpdate(isLeftAgent bool) {
 
+// LeaveUpdateDivide is used for updating the other agent during the divide algorithm
+func (agent *Agent) LeaveUpdateDivide() {
+	oppositeDirection := (agent.Direction + 1) % (2)
```
pourquoi des parenthèses autour de 2? 

`(2)`

SeanLF replied:
j'ai oublié de les enlever quand j'ai remplacé l'expression par `2`

SeanLF replied:
#7 ✅

jonathanGB commented on main.go:
```diff
@@ -165,3 +168,63 @@ func optTeamSize(ring bhs.Ring) (blackHoleNodeID uint64) {
+		go func(direction bhs.Direction, ch chan<- bool, blackhole chan<- uint64) {
+			agent := bhs.NewAgent(direction, ring, cautiousWalk)
+			agent.ActAsSmall = false // for update catching
+			agent.UnexploredSet = [2]uint64{1, uint64(len(agent.Ring) - 1)}
```
serait peut-être mieux de définir un type pour le set au lieu d'utiliser `[2]uint64`.

Suggestion: `type ExplorationSet [2]uint64`

SeanLF replied:
et peut être créer un type pour remplacer uint64 (`NodeID`?)

jonathanGB replied:
+1

SeanLF replied:
voir #7

jonathanGB commented on bhs/agent.go:
```diff
@@ -206,6 +206,20 @@ func (agent *Agent) LeaveUpdate(isLeftAgent bool) {
+	oppositeDirection := (agent.Direction + 1) % (2)
+	agent.MoveToLastExplored(oppositeDirection)
+
+	whiteboard := agent.Position.whiteboard
```
je me demande s'il faut pas plutôt prendre l'adresse du whiteboard

`whiteboard := &(agent.Position.whiteboard)`

Sinon, je crois qu'on copie le whiteboard. Faudrait vérifier.

SeanLF replied:
`agent.Position.whiteboard` est du type `*Whiteboard`, donc on reçoit un pointeur vers le même objet
👍?

SeanLF committed:
Turns out I lied

can now find the black hole anywhere

SeanLF committed:
simplification

SeanLF committed:
Fix an oopsie

Middle group destination was wrong

SeanLF committed:
Cleanup

SeanLF committed:
Add benchmark

SeanLF committed:
refactor (#7)

- use NodeID instead of uint64 where possible (this lead to some necessary typecasting for type correctness)
- renamed some variables to be more accurate
- added a few comments
- changed cautiousWalk to a constant where appropriate

jonathanGB committed:
Merge pull request #6 from jonathanGB/group

Implement Group
